File: src/main/java/me/oriharel/machinery/inventory/InventoryNavigationItemData.java
package me.oriharel.machinery.inventory;

import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

/**
 * Requires an InventoryPage that implements DatablePage
 * @param <T> type of data
 */
public class InventoryNavigationItemData<T> extends InventoryNavigationItem implements Datable<T> {

    private T navigationData;

    public InventoryNavigationItemData(String routeToName, Inventory parentInventory, int indexInInventory, Material material, int amount, String displayName,
                                       T navigationData, String... lore) {
        super(routeToName, parentInventory, indexInInventory, material, amount, displayName, lore);
        this.navigationData = navigationData;
    }

    public InventoryNavigationItemData(String routeToName, Inventory parentInventory, int indexInInventory, ItemStack itemStack, T navigationData) {
        super(routeToName, parentInventory, indexInInventory, itemStack);
        this.navigationData = navigationData;
    }

    public InventoryNavigationItemData(String routeToName, Inventory parentInventory, int indexInInventory, Material material, int amount, String displayName,
                                       T navigationData) {
        super(routeToName, parentInventory, indexInInventory, material, amount, displayName);
        this.navigationData = navigationData;
    }

    public T getNavigationData() {
        return navigationData;
    }

    @Override
    public InventoryNavigationItemData<T> navigate() {
        parentInventory.navigateToNamedRoute(routeToName, navigationData);
        return this;
    }

    @Override
    public T getStoredData() {
        return navigationData;
    }

    @Override
    public InventoryNavigationItemData<T> setStoredData(T data) {
        this.navigationData = data;
        return this;
    }
}

File: src/main/java/me/oriharel/machinery/inventory/InventoryNavigationItemData.kt
package me.oriharel.machinery.inventory

import org.bukkit.Material
import org.bukkit.inventory.ItemStack

/**
 * Requires an InventoryPage that implements DatablePage
 * @param <T> type of data
</T> */
class InventoryNavigationItemData<T> : InventoryNavigationItem, NavigableData<T> {
    override var storedData: T
        private set
        public get() {
            return field
        }
    set
    private var routePage: NavigableDataInventoryPage<T>

    constructor(routePage: NavigableDataInventoryPage<T>, parentInventory: Inventory, indexInInventory: Int, material: Material?, amount: Int, displayName: String?,
                navigationData: T, vararg lore: String?) : super(null, parentInventory, indexInInventory, material, amount, displayName, *lore) {
        storedData = navigationData
        this.routePage = routePage
    }

    constructor(routePage: NavigableDataInventoryPage<T>, parentInventory: Inventory, indexInInventory: Int, itemStack: ItemStack?, navigationData: T) : super(null, parentInventory, indexInInventory, itemStack) {
        storedData = navigationData
        this.routePage = routePage
    }

    constructor(routePage: NavigableDataInventoryPage<T>, parentInventory: Inventory, indexInInventory: Int, material: Material?, amount: Int, displayName: String?,
                navigationData: T) : super(null, parentInventory, indexInInventory, material, amount, displayName) {
        storedData = navigationData
        this.routePage = routePage
    }

    override fun navigate(): InventoryNavigationItemData<T> {
        parentInventory.navigateDirect(routePage, storedData)
        return this
    }

    override fun setStoredData(data: T): InventoryNavigationItemData<T> {
        storedData = data
        return this
    }
}